### PR TITLE
[gosrc2cpg] Formation of fully qualified namespace from package name and Initial base AST nodes along with namespace blocks.

### PR DESCRIPTION
**Pull request by pandurangpatil**
**Formation of fully qualified namespace from package name**

1. Updated goastparser version to pickup the changes to read go.mod file metadata.
2. Made changes to separately identify the `go.mod.json` file
3. Used circa library to unmarshal JSON to scala entity to read `.mod.json`
4. Started added NamespaceBlock unit test, with one working unit test
5. Prepared the utility which will form the namespace based on the information available along with respective unit tests covering majority of the possibilities.

a. If there is no `go.mod` file and `<root project path>/test.go` (package `main`) => `main`
b. `<root project path>/first/second/test.go` (package `main`) => `/first/second/main`
c. `<root project path>/test.go` (package `main`) => `main`
d. `go.mod` (module `jorn.io/trial`) and `<root project path>/foo.go`(package `trial`) => `jorn.io/trial`
e. `go.mod` (module `jorn.io/trial`) and `<root project path>/foo.go` (package `foo`) => `jorn.io/trial>foo`
f. `go.mod` (module `jorn.io/trial`) and `<root project path>/first/foo.go` (package `first`) => `jorn.io/trial/first`
g. `go.mod` (module `jorn.io/trial`) and `<root project path>/first/foo.go` (package `bar`) => `jorn.io/trial/first>bar`

if you pay attention to `e` and `g` use case where package name is not matching with the wrapped folder. Then in that case go makes use of alias mechanism like

If I consider `g` use case and try to import that package somewhere else I will be doing it as follows

`import bar "jorn.io/trial/first"`. I am representing this alias as `jorn.io/trial/first>bar` to form the namespace. If anyone has any different suggestions add your comments.

**Initial base AST nodes along with namespace blocks.**
1. Incorporated initial root AST nodes using the package full name generated using the utility.
2. Added utility method in `x2cpg` to create `method return node`. Have added a respective comment in c2cpg where I tried to use the method from `x2cpg` and tried removing it from `c2pg`. However it occurred to me, it has been getting used from multiple places in `c2cpg` module. Hence I left with TODO
3. Addressed all the review comments by David.
4. There is still scope for few more unit tests, I will try to cover them in next PR.

### CODE REVIEW

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    val config = Config()
+    config.inputPath = Paths.get("").toAbsolutePath.toString + "/"
+    GoMod.config = config
+    GoMod.meta = GoMod(
+      fileFullPath = Paths.get(config.inputPath, "go.mod").toAbsolutePath.toString,
+      module = GoModModule("joern.io/trial"),
+      dependencies = List[GoModDependency]()
+    )
+    val namespace =
+      GoMod.getNameSpace(Paths.get(config.inputPath, "first", "second", "test.go").toAbsolutePath.toString, "main")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala:56]: With `better.io.Files` you can also use
```scala
import better.files.File
import scala.language.postfixOps

GoMod.getNameSpace(File(config.input) / "first" / "second" / "test.go" pathAsString, "main")
```

Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala:56]: Something interesting to learn, I appreciate the suggestions :)

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      return tokens.mkString("/")
+    }
+
+    // go.mod (module jorn.io/trial) and <root project path>/foo.go (package trial) => jorn.io/trial
+    // go.mod (module jorn.io/trial) and <root project path>/foo.go (package foo) => jorn.io/trial>foo
+    // go.mod (module jorn.io/trial) and <root project path>/first/foo.go (package first) => jorn.io/trial/first
+    // go.mod (module jorn.io/trial) and <root project path>/first/foo.go (package bar) => jorn.io/trial/first>bar
+    val remainingpath              = compilationUnitFilePath.stripPrefix(config.inputPath)
+    val pathTokens                 = remainingpath.split(File.separator)
+    val tokens: ListBuffer[String] = ListBuffer.empty[String]
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      // In order to isolate the main package from the root folder main package. We are using folder structure from
+      // root project path appended with main
+      //
+      // e.g.
+      // 1. if there is go file inside <root project path>/first/second/test.go (package main) => '/first/second/main'
+      // 2. <root project path>/test.go (package main) => 'main'
+
+      val remainingpath = compilationUnitFilePath.stripPrefix(config.inputPath)
+      import java.util.regex.Pattern
+      val pattern                    = Pattern.quote(System.getProperty("file.separator"))
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    nameSpace
+  }
+}
+
+case class GoMod(fileFullPath: String, module: GoModModule, dependencies: List[GoModDependency])
+case class GoModModule(name: String, lineNo: Int = 0, colNo: Int = 0, endLineNo: Int = 0, endColNo: Int = 0)
+case class GoModDependency(
+  module: String,
+  version: String,
+  lineNo: Int = 0,
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/model/GoMod.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala [hunk 5]
```diff
@@ -17,4 +21,15 @@ object GoAstJsonParser {
...
   }
+
+  def readModFile(file: Path): GoMod = {
+    val jsonContent = IOUtils.readLinesInFile(file).mkString
+    decode[GoMod](jsonContent) match {
+      case Right(myData) =>
+        myData
+      case Left(error) =>
+        logger.warn(s"Error decoding JSON - '${file.toString}': $error")
+        null
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/utils/AstGenRunner.scala [hunk 6]
```diff
@@ -14,6 +14,7 @@ import scala.util.{Failure, Success, Try}
 object AstGenRunner {
   private val logger = LoggerFactory.getLogger(getClass)
   case class AstGenRunnerResult(
+    parsedModFile: (String, String) = ("", ""),
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/utils/AstGenRunner.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    namespace shouldBe "first/second/main"
+  }
+
+  "with .mod file and main pkg 2 use case" in {
+    val config = Config()
+    config.inputPath = Paths.get("").toAbsolutePath.toString + "/"
```
[the remark this answers is not in the document]
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala (hunk 1 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/model/GoModTest.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala [hunk 8]
```diff
@@ -20,6 +24,13 @@ class GoSrc2Cpg extends X2CpgFrontend[Config] {
       File.usingTemporaryDirectory("gosrc2cpgOut") { tmpDir =>
         new MetaDataPass(cpg, Languages.GOLANG, config.inputPath).createAndApply()
         val astGenResult = new AstGenRunner(config).execute(tmpDir)
+        GoMod.config = Some(config)
+        (if (!astGenResult.parsedModFile.isEmpty)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala:28]: Could this code not simply be: 
```scala
astGenResult.parsedModFile.foreach(modFile => 
    GoAstJsonParser.readModFile(Paths.get(modFile)).foreach(x => GoMod.meta = x)
)

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala [hunk 9]
```diff
@@ -26,14 +27,44 @@ class AstCreator(val config: Config, val parserResult: ParserResult)
 
   override def createAst(): DiffGraphBuilder = {
 
-    val fileNode = NewFile().name(parserResult.filename).order(1)
-
-    val astForDeclaration = parserResult.json(ParserKeys.Decls).arr.map(item => astForNode(item))
-    val ast               = Ast(fileNode).withChildren(astForDeclaration)
+//    val fileNode = NewFile().name(parserResult.filename).order(1)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala:30]: Remove commented out code

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala [hunk 10]
```diff
@@ -26,14 +27,44 @@ class AstCreator(val config: Config, val parserResult: ParserResult)
...
     Ast.storeInDiffGraph(ast, diffGraph)
     diffGraph
   }
 
+  private def astForTranslationUnit(rootNode: ParserNodeInfo): Ast = {
+    val fullQualifiedPackage =
+      GoMod.getNameSpace(parserResult.filename, parserResult.json(ParserKeys.Name)(ParserKeys.Name).str)
+    val namespaceBlock = NewNamespaceBlock()
+      .name(fullQualifiedPackage)
+      .fullName(s"$relPathFileName:${fullQualifiedPackage}")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala:42]:
```suggestion
      .fullName(s"$relPathFileName:$fullQualifiedPackage")
```

Comment (DavidBakerEffendi) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala:42]: Also rename `fullQualifiedPackage` to `fullyQualifiedPackage`

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala [hunk 11]
```diff
@@ -17,4 +23,15 @@ object GoAstJsonParser {
     val sourceFileContent = IOUtils.readEntireFile(filePath)
     ParserResult(filePath.getFileName.toString, fullFilePath, json, sourceFileContent)
   }
+
+  def readModFile(file: Path): Option[GoMod] = {
+    val jsonContent = IOUtils.readLinesInFile(file).mkString
+    val t = decode[GoMod](jsonContent) match {
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala:29]: The `t` variable is unnecessary here
```scala
    decode[GoMod](jsonContent) match {
       case Left(error) =>
         logger.warn(s"Error decoding JSON - '${file.toString}': $error")
         Left(error)
       case x => x
     }.toOption
```

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/NamespaceBlockTests.scala [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    val List(x) = cpg.namespaceBlock.filenameNot(FileTraversal.UNKNOWN).l
+    x.name shouldBe "main"
+    x.filename should not be empty
+    x.fullName shouldBe s"Test0.go:main"
+    x.order shouldBe 1
+  }
+
+  // TODO: Once the method node creation is done, this unit test needs to be fixed to add fullname for "func foo()" as well, along with dummy method representing the file.
+  "should allow traversing from namespace block to method" ignore {
+    cpg.namespaceBlock.filenameNot(FileTraversal.UNKNOWN).ast.isMethod.fullName.l shouldBe List(s"Test0.go:main")
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/NamespaceBlockTests.scala:31]: No interpolation is being done so the `s` can be removed
```suggestion
    cpg.namespaceBlock.filenameNot(FileTraversal.UNKNOWN).ast.isMethod.fullName.l shouldBe List("Test0.go:main")
```

**Review by DavidBakerEffendi — COMMENTED** [review]
Minor suggestions, should be good to merge after that

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala (hunk 8 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala:28]: There can be only one go.mod file present. I guess I could simplify it as following.

```
        GoMod.meta =
          if (!astGenResult.parsedModFile.isEmpty)
            GoAstJsonParser.readModFile(Paths.get(astGenResult.parsedModFile.get))
          else None
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala (hunk 11 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala:29]: For some reason, `.toOption` is throwing a compilation error.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala (hunk 9 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala (hunk 10 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstCreator.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/NamespaceBlockTests.scala (hunk 12 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/test/scala/io/joern/go2cpg/passes/ast/NamespaceBlockTests.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala (hunk 11 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala:29]: Does the following perhaps work
```scala
    (decode[GoMod](jsonContent) match {
       case Left(error) =>
         logger.warn(s"Error decoding JSON - '${file.toString}': $error")
         Left(error)
       case x => x
     }).toOption
 ```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala (hunk 8 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala]: The `Option` class implements `foreach` as a means to operate on the value of it exists, so no need to do a `isDefined` check or `get` call.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala (hunk 11 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/parser/GoAstJsonParser.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala (hunk 8 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/GoSrc2Cpg.scala]: done

**Review by DavidBakerEffendi — APPROVED** [review]

**Comment by mpollmeier** [comment]
small fixup for versioning: https://github.com/joernio/joern/pull/2980

**Pull request merged**